internal/telemetry/region_pusher.go

```diff
@@ -0,0 +1,147 @@
+		logger:    logger,
+		instance:  instance,
+		regionID:  regionID,
+		telemetry: make(map[int64]map[sm.CheckClass]*sm.CheckClassTelemetry),
+	}
+
+	var ticker ticker = newStdTicker(timeSpan)
+
+	for _, o := range opts {
+		switch o := o.(type) { //nolint:gocritic
+		case withTicker:
+			ticker = o
+		}
+	}
+
+	go tp.run(ctx, ticker)
+
+	return tp
+}
+
+type withTicker ticker
+
+func (p *RegionPusher) run(ctx context.Context, ticker ticker) {
+	p.logger.Info().Msg("region pusher starting")
+
+	// TODO: We could potentially create here a pushCtx, pass that to push, and
+	// from push keep retrying sending the data if it fails initially until the
+	// pushCtx is canceled, which should happen once the ticker ticks again, to
+	// avoid overlapping push requests.
+	// By now, only retry folllowing the ticker's pace.
+
+LOOP:
+	for {
+		select {
+		case <-ticker.C():
+			p.logger.Info().Msg("pushing telemetry")
+			m := p.next()
+			go p.push(m) // Avoid blocking
+		case <-ctx.Done():
+			p.logger.Debug().Msg("region pusher stopping")
+			m := p.next()
+			p.push(m)
+			break LOOP
+		}
+	}
+}
+
+// AddExecution adds a new execution to the tenant telemetry.
+func (p *RegionPusher) AddExecution(e Execution) {
+	p.telemetryMu.Lock()
+	defer p.telemetryMu.Unlock()
+
+	tenantTele, ok := p.telemetry[e.LocalTenantID]
+	if !ok {
+		tenantTele = make(map[sm.CheckClass]*sm.CheckClassTelemetry)
+		p.telemetry[e.LocalTenantID] = tenantTele
+	}
+
+	clTele, ok := tenantTele[e.CheckClass]
+	if !ok {
+		clTele = &sm.CheckClassTelemetry{CheckClass: e.CheckClass}
+		tenantTele[e.CheckClass] = clTele
+	}
+
+	clTele.Executions++
+	clTele.Duration += float32(e.Duration.Seconds())
+}
+
+func (p *RegionPusher) next() sm.RegionTelemetry {
+	m := sm.RegionTelemetry{
+		Instance:  p.instance,
+		RegionId:  p.regionID,
+		Telemetry: make([]*sm.TenantTelemetry, 0, len(p.telemetry)),
+	}
+
+	p.telemetryMu.Lock()
+	defer p.telemetryMu.Unlock()
+
+	// Copy current telemetry data
+	for tenantID, tTele := range p.telemetry {
+		tenantTele := &sm.TenantTelemetry{
+			TenantId:  tenantID,
+			Telemetry: make([]*sm.CheckClassTelemetry, 0, len(tTele)),
+		}
+		for _, clTele := range tTele {
+			tenantTele.Telemetry = append(tenantTele.Telemetry, &sm.CheckClassTelemetry{
+				CheckClass: clTele.CheckClass,
+				Executions: clTele.Executions,
+				Duration:   clTele.Duration,
+			})
+		}
+		m.Telemetry = append(m.Telemetry, tenantTele)
+	}
+
+	return m
+}
+
+func (p *RegionPusher) push(m sm.RegionTelemetry) {
+	// We don't want to cancel a possibly ongoing request even if the agent
+	// context is done, therefore use background context
+	ctx := context.Background()
+	r, err := p.client.PushTelemetry(ctx, &m)
+	if err != nil {
+		p.logger.Err(err).Msg("error pushing telemetry")
+		return
+	}
+	if r.Status.Code != sm.StatusCode_OK {
+		p.logger.Error().
+			Int32("statusCode", int32(r.Status.Code)).
+			Str("statusMessage", r.Status.Message).
+			Msg("error pushing telemetry")
+	}
+}
```

Comment:
this might benefit from some observability:

- keep track of the number of pushes
- keep track of the duration of those pushes
- keep track of the number of errors

Reply:
Right, I hesitated to add it because this is being measured from the proxy and this one is executed within the same cluster and namespace, but it makes sense to measure it from the agent side. See [f622450](https://github.com/grafana/synthetic-monitoring-agent/pull/621/commits/f62245037d2b93878abc6418f3a4a6d87e96e4b6).